pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,127 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsSourceReady is the heartbeat endpoint for user defined source gRPC.
+  rpc IsSourceReady(google.protobuf.Empty) returns (SourceReadyResponse);
+}
+
+/*
+ * EventTime represents the event time of the datum.
+ */
+message EventTime {
+  google.protobuf.Timestamp event_time = 1;
+}
+
+/*
+ * ReadRequest is the request for reading datum stream from user defined source.
+ */
+message ReadRequest {
+  message Request {
+    // num_records is the number of records to read.
+    uint64 num_records = 1;
+  }
+  Request request = 1;
+}
+
+/*
+ * ReadResponse represents a list of datum response elements.
+ */
+message ReadResponse {
+  message Result {
+    // elements is a list of datum response elements.
+    repeated DatumResponse elements = 1;
+  }
+  Result result = 1;
+}
+
+/*
+ * DatumResponse represents a datum response element.
+ */
+message DatumResponse {
```

Comment:
I see that we have many [nested types](https://protobuf.dev/programming-guides/proto3/#nested) in this design.. looks like it's used if we have an inner type needs to be `repeated` or just to group the types together... (in my understanding..)

in that case... would we consider moving this type inside line 48?

Reply:
@jy4096 I don't have a good reference for when to use nested types in general. The way I apply nested type is that if I am sure type A is going to be ONLY used by type B, I put type A as an inner type of type B.

Hence you can see I apply the nested type approach mostly for the `Response` and `Request` messages. Regarding `DatumResponse`, currently, it's only used by `ReadResponse`, however I didn't nest it because I want to leave the door open for future use cases. e.g. we might want to add a new API ReadFromPartition(int partitionID) which reads from a specified ID and return a list of `DatumResponse`.